homeassistant/components/integration/sensor.py

```diff
@@ -253,6 +343,33 @@ def _unit(self, source_unit: str) -> str:
 
         return self._unit_template.format(integral_unit)
 
+    def _derive_and_set_attributes_from_state(self, source_state: State) -> None:
+        unit = source_state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
+        if unit is not None:
+            self._unit_of_measurement = self._unit(unit)
+
+        if (
+            self.device_class is None
+            and source_state.attributes.get(ATTR_DEVICE_CLASS)
+            == SensorDeviceClass.POWER
+        ):
+            self._attr_device_class = SensorDeviceClass.ENERGY
+            self._attr_icon = None
+
+        self.async_write_ha_state()
+
+    def _update_integral(self, area: Decimal) -> None:
+        area_scaled = area / (self._unit_prefix * self._unit_time)
+        if isinstance(self._state, Decimal):
+            self._state += area_scaled
+        else:
+            self._state = area_scaled
+        _LOGGER.debug(
+            "area = %s, area_scaled = %s new state = %s", area, area_scaled, self._state
+        )
+        self._last_valid_state = self._state
+        self.async_write_ha_state()
```

Comment:
This might be an issue, as the state might be written twice ( _derive_and_set_attributes_from_state plus this function)

Reply:
I've removed the write state call from `_derive_and_set_attributes_from_state`. Then now it will only be called once: on unavailability of the source or successful integration.

Note however that this call was also made twice during the state change `calc_integration` method.